Improvements to the display of the experience, todo and task charts

ltorroba opened the pull request:
1) Fixes issue #6679, which referred to the EXP chart getting cropped.
2) Added some bottom padding to the EXP chart so that the tags are
further apart from the chart itself (requires rebuilding stylus)
3) Added event handlers for the dynamic resizing of the charts. The
event handlers use the same data every time and are only created when a
chart is first toggled, as to reduce overhead.
4) Moved the location of the tasks chart into the .task-text, so that
the habit +/- buttons are now fully triggerable.

I’ll add more details on the implementation below soon.

ltorroba committed:
1) Fixes issue #6679, which referred to the EXP chart getting cropped.
2) Added some bottom padding to the EXP chart so that the tags are
further apart from the chart itself (requires rebuilding stylus)
3) Added event handlers for the dynamic resizing of the charts. The
event handlers use the same data every time and are only created when a
chart is first toggled, as to reduce overhead.
4) Moved the location of the tasks chart into the .task-text, so that
the habit +/- buttons are now fully triggerable.

ltorroba commented:
I picked up issue #6679, and noticed that alongside fixing the clipping problem this would be a good chance to fix a few other bugs related to charts. They were listed above. On the whole, I believe this PR would make the chart somewhat prettier, especially when it comes to dynamic/flexible container sizes.

## Charts currently at desktop resolution (Safari @ Mac):

### EXP

![](http://i.imgur.com/59777HY.png)

### Todos

![](http://i.imgur.com/TFCpjuq.png)

### Task

Portion of the `+` sign becomes unclickable as the chart floats above it, See next section for the specific region,
![](http://i.imgur.com/KhCRsuP.png)
_The same behaviour was observed on Chrome on the original issue, and confirmed by tests._

## Charts currently, at reduced resolution (Chrome @ Mac):

### EXP

![](http://i.imgur.com/qjl9PuJ.png)

### Todos

![](http://i.imgur.com/J9KRXgy.png)

### Task

_Moreover, bounded region becomes unclickable_
![](http://i.imgur.com/pRGPuQ4.png)
_I also verified this on Firefox and Safari._

## Charts after these commits, at desktop resolution (Chrome @ Mac)

### EXP

![](http://i.imgur.com/migI2Tt.png)

### Todos

![](http://i.imgur.com/mle6w42.png)

### Task

![](http://i.imgur.com/LMFCC82.png)
_Also tested on Firefox and Safari._

## Charts after these commits, at reduced resolution (Chrome @ Mac)

### EXP

![](http://i.imgur.com/5iU0Ai2.png)

### Todos

![](http://i.imgur.com/GeVJEXI.png)

### Task

![](http://i.imgur.com/73zdepZ.png)

# The only problem

Only exception is at very specific screen sizes (e.g.: task column div width is between 271 and 273 pixels) on Safari and Firefox, where the following may occur:
![](http://i.imgur.com/yVrc89l.png)

I have tried different workarounds but it seems to be a library and browser issue, so we are not talking anything neat. Any ideas? Should we fix this, and/or try to report it as an issue with the Google Charts API itself?

keysmashes commented:
Is it even worth showing the legend? It seems like it should be self-evident, and 'score' isn't very informative anyway.

crookedneighbor commented on website/public/js/controllers/rootCtrl.js:
```diff
@@ -169,7 +169,13 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
     $rootScope.toggleChart = function(id, task) {
-      var history = [], matrix, data, chart, options;
+      if($rootScope.charts[id] === undefined) {
+        $(window).resize(function() { 
```
Window resize functions should be debounced.

crookedneighbor commented on website/public/js/controllers/rootCtrl.js:
```diff
@@ -169,7 +169,13 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
     $rootScope.charts = {};
     $rootScope.toggleChart = function(id, task) {
-      var history = [], matrix, data, chart, options;
+      if($rootScope.charts[id] === undefined) {
```
A simple `if (!$rootScope.charts[id])` should suffice here.

crookedneighbor commented on website/views/options/social/tavern.jade:
```diff
@@ -83,7 +83,7 @@
                 div
                   p
-                    span.achievement.achievement-firefox
+                    span.achievement.achievement-boot
```
This doesn't look related to this PR.

crookedneighbor commented on website/public/js/controllers/rootCtrl.js:
```diff
@@ -184,23 +190,62 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
+      $rootScope.drawChart(id, data);
+    };
+
+    $rootScope.drawChart = function(id, data, options) {
```
Does this need to be exposed onto `$rootScope`? Looks like it is only called internally.

ltorroba committed:
Removed legend

- Removed the legend as suggested by @anowlcalledjosh.
- Made small adjustments on the widths of the charts depending on their
environments so that they are properly aligned. This reflects the
changes following the removal of the legend.

ltorroba committed:
Assorted improvements

Implemented suggestions made by @crookedneighbor:
- If notation simplified
- `drawChart` is no longer exposed to the `rootScope`
- Debounced window resize function

Moreover:
- Fixed bug with the resize function that caused it not to trigger when
first toggling a chart

ltorroba commented:
@anowlcalledjosh @crookedneighbor Sorry for the delay, but here are the commits associated with your suggestions. Let me know if there's anything else to be done. 😄

Alys commented:
@zesme Thanks for that! Just so you know, we won't be able to review this until late this month because we have a code freeze now while we finish the final stages of version 3 of Habitica's API. If you'd like more information, see the new "API Changes" section in Guidance for Blacksmiths at http://habitica.wikia.com/wiki/Guidance_for_Blacksmiths#API_Changes and you're welcome to ask questions in the Aspiring Coders guild.

ltorroba commented:
@Alys No worries, I had heard of the code freeze but, ironically, this To-Do had gotten quite red in my list, so I felt it was time to get it over with. As soon as the new API gets released I will get working on the merging.

awitherow commented on website/views/shared/tasks/task.jade:
```diff
@@ -14,4 +14,4 @@ li(id='task-{{::task.id}}',
     include ./edit/index
 
-  div(class='{{obj._id}}{{task.id}}-chart', ng-show='charts[obj._id+task.id]')
+    
```
is this necessary?

ltorroba replied:
Good point! Not quite sure what was going on in my head when I removed that. I just made a push to bring it back.

awitherow commented on website/public/js/controllers/rootCtrl.js:
```diff
@@ -184,23 +184,73 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
           if (task && task._editing) task._editing = false;
       }
+
+      if($rootScope.charts[id]) {
```
I see a lot of cases with spaces between the if and the parenthesis, and also a lot of cases not with. Is there a code styleguide that is being followed?

crookedneighbor replied:
We've instituted eslint for the server code. We're about to do a rewrite of the client code, so we haven't bothered including it in the linting, since it will be replaced soon.

ltorroba commented:
Fixed conflicts with base branch

crookedneighbor commented on website/client/js/controllers/rootCtrl.js:
```diff
@@ -185,23 +185,73 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
+          drawChart(id, data);
+        }, 300);
+
+        $(window).resize(handleResize);
```
This is adding a new event listener to the window resize every time the chart is toggled. Meaning if you toggle back and forth a few times, you add a new listener each time and the old listeners do not get cleaned up. This will be a performance issue.

crookedneighbor commented on website/client/js/controllers/rootCtrl.js:
```diff
@@ -185,23 +185,73 @@ habitrpg.controller("RootCtrl", ['$scope', '$rootScope', '$location', 'User', '$
+    function drawChart(id, data, options) {
+      var chart, width;
+
+      switch(id) {
```
Let's use `if/else if/else` here instead of a switch.

crookedneighbor commented:
Sorry for the super long delay on reviewing this. It looks 95% of the way there. The one big problem with it is that it's setting window resize listeners without cleaning up after itself.

ltorroba commented:
@crookedneighbor No worries! I am kinda busy this week, but as soon as I get a chance I will get everything sorted.

crookedneighbor commented:
@zesme Status report on this?

ltorroba commented:
@crookedneighbor I'll be taking a look at this today or within the next two days by the latest. Sorry about the delay!

ltorroba committed:
fix(rootCtrl): fix duplicate window resize events

Window resize event handlers are were being duplicated. This was fixed
by creating another variable, `$rootScope.resizeCharts`, that holds the
resize handlers for each one of the charts.

Better solution ideas are welcome.

ltorroba committed:
style(rootCtrl): remove debug message

ltorroba committed:
style(rootCtrl): change switch for if/else if/else

ltorroba commented:
@crookedneighbor Made the changes you asked for. 66922a6 seems kinda hacky to me, but I couldn't figure out a better way to implement it. If you (or anyone else) has any ideas I'll happily improve it!

crookedneighbor commented:
Did a tiny adjustment: e3b9636c42d2641fe038df6226ad548e88d4f1b3

Basically, just set one listener on the window and looped through the resize functions for each active chart.

Notated toward your next tier.
